File: applications/rtksubselect/rtksubselect.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtksubselect_ggo.h"
#include "rtkGgoFunctions.h"

#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkProjectionsReader.h"
#include "rtkConstantImageSource.h"

#include <itkImageFileWriter.h>
#include <itkRegularExpressionSeriesFileNames.h>
#include <itkPasteImageFilter.h>

int
main(int argc, char * argv[])
{
  GGO(rtksubselect, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;
  using OutputImageType = itk::Image<OutputPixelType, Dimension>;

  // Projections reader
  using ReaderType = rtk::ProjectionsReader<OutputImageType>;
  auto reader = ReaderType::New();
  rtk::SetProjectionsReaderFromGgo<ReaderType, args_info_rtksubselect>(reader, args_info);
  TRY_AND_EXIT_ON_ITK_EXCEPTION(reader->Update())

  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::endl;
  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));

  // Compute the indices of the selected projections
  std::vector<int> indices;
  int              n = geometry->GetGantryAngles().size();
  if (args_info.last_given)
    n = std::min(args_info.last_arg, n);
  if (args_info.list_given)
    for (unsigned int i = 0; i < args_info.list_given; i++)
    {
      indices.push_back(args_info.list_arg[i]);
    }
  else
    for (int noProj = args_info.first_arg; noProj < n; noProj += args_info.step_arg)
    {
      indices.push_back(noProj);
    }

  // Output RTK geometry object
  using GeometryType = rtk::ThreeDCircularProjectionGeometry;
  auto outputGeometry = GeometryType::New();

  // Output projections object
  using SourceType = rtk::ConstantImageSource<OutputImageType>;
  auto source = SourceType::New();
  source->SetInformationFromImage(reader->GetOutput());
  OutputImageType::SizeType outputSize = reader->GetOutput()->GetLargestPossibleRegion().GetSize();
  outputSize[Dimension - 1] = indices.size();
  source->SetSize(outputSize);
  source->SetConstant(0);
  TRY_AND_EXIT_ON_ITK_EXCEPTION(source->Update())

  // Fill in the outputGeometry and the output projections
  using PasteType = itk::PasteImageFilter<OutputImageType>;
  auto paste = PasteType::New();
  paste->SetSourceImage(reader->GetOutput());
  paste->SetDestinationImage(source->GetOutput());

  OutputImageType::RegionType sourceRegion;
  OutputImageType::IndexType  destinationIndex;
  for (unsigned int i = 0; i < indices.size(); i++)
  {
    // If it is not the first projection, we need to use the output of
    // the paste filter as input
    if (i)
    {
      OutputImageType::Pointer pimg = paste->GetOutput();
      pimg->DisconnectPipeline();
      paste->SetDestinationImage(pimg);
    }

    sourceRegion = reader->GetOutput()->GetLargestPossibleRegion();
    sourceRegion.SetIndex(Dimension - 1, indices[i]);
    sourceRegion.SetSize(Dimension - 1, 1);
    paste->SetSourceRegion(sourceRegion);

    destinationIndex = reader->GetOutput()->GetLargestPossibleRegion().GetIndex();
    destinationIndex[Dimension - 1] = i;
    paste->SetDestinationIndex(destinationIndex);

    TRY_AND_EXIT_ON_ITK_EXCEPTION(paste->Update())

    // Fill in the output geometry object
    outputGeometry->SetRadiusCylindricalDetector(geometry->GetRadiusCylindricalDetector());
    outputGeometry->AddProjectionInRadians(geometry->GetSourceToIsocenterDistances()[indices[i]],
                                           geometry->GetSourceToDetectorDistances()[indices[i]],
                                           geometry->GetGantryAngles()[indices[i]],
                                           geometry->GetProjectionOffsetsX()[indices[i]],
                                           geometry->GetProjectionOffsetsY()[indices[i]],
                                           geometry->GetOutOfPlaneAngles()[indices[i]],
                                           geometry->GetInPlaneAngles()[indices[i]],
                                           geometry->GetSourceOffsetsX()[indices[i]],
                                           geometry->GetSourceOffsetsY()[indices[i]]);
    outputGeometry->SetCollimationOfLastProjection(geometry->GetCollimationUInf()[indices[i]],
                                                   geometry->GetCollimationUSup()[indices[i]],
                                                   geometry->GetCollimationVInf()[indices[i]],
                                                   geometry->GetCollimationVSup()[indices[i]]);
  }

  // Geometry writer
  if (args_info.verbose_flag)
    std::cout << "Writing geometry information in " << args_info.out_geometry_arg << "..." << std::endl;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(rtk::WriteGeometry(outputGeometry, args_info.out_geometry_arg))

  // Write
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(paste->GetOutput(), args_info.out_proj_arg))

  return EXIT_SUCCESS;
}

File: applications/rtksubselect/rtksubselect.py
#!/usr/bin/env python
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    # argument parsing
    parser = rtk.RTKArgumentParser(
        description="Subselect projections from a stack and write updated geometry."
    )

    parser.add_argument(
        "--geometry", "-g", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--out_geometry", help="Output geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--out_proj", help="Output projections stack file name", type=str, required=True
    )

    parser.add_argument(
        "--first", "-f", help="First projection index", type=int, default=0
    )
    parser.add_argument("--last", "-l", help="Last projection index", type=int)
    parser.add_argument(
        "--step", "-s", help="Step between projections", type=int, default=1
    )
    parser.add_argument(
        "--list",
        help="List of projection indices to keep (0-based)",
        type=int,
        nargs="+",
    )
    # RTK specific groups
    rtk.add_rtkinputprojections_group(parser)

    # Parse the command line arguments
    return parser


def process(args_info: argparse.Namespace):
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Projections reader
    reader = rtk.ProjectionsReader[OutputImageType].New()
    rtk.SetProjectionsReaderFromArgParse(reader, args_info)

    # Geometry
    if args_info.verbose:
        print(f"Reading geometry information from {args_info.geometry}...")
    geometry = rtk.read_geometry(args_info.geometry)

    # Compute the indices of the selected projections
    indices = []
    n = len(geometry.GetGantryAngles())
    if args_info.last:
        n = min(args_info.last, n)
    if args_info.list:
        for i in args_info.list:
            indices.append(i)
    else:
        start = args_info.first
        step = args_info.step
        for noProj in range(start, n, step):
            indices.append(noProj)

    # Output RTK geometry object
    outputGeometry = rtk.ThreeDCircularProjectionGeometry.New()

    # Output projections object
    source = rtk.ConstantImageSource[OutputImageType].New()
    source.SetInformationFromImage(reader.GetOutput())
    outputSize = itk.size(reader.GetOutput())
    outputSize_list = [s for s in outputSize]
    outputSize_list[Dimension - 1] = len(indices)
    source.SetSize(outputSize_list)
    source.SetConstant(0)
    source.Update()

    # Fill in the outputGeometry and the output projections
    paste = itk.PasteImageFilter[OutputImageType].New()
    paste.SetSourceImage(reader.GetOutput())
    paste.SetDestinationImage(source.GetOutput())

    for i, src_idx in enumerate(indices):
        # If it is not the first projection, we need to use the output of the paste filter as input
        if i:
            pimg = paste.GetOutput()
            pimg.DisconnectPipeline()
            paste.SetDestinationImage(pimg)

        sourceRegion = reader.GetOutput().GetLargestPossibleRegion()
        sourceRegion.SetIndex(Dimension - 1, src_idx)
        sourceRegion.SetSize(Dimension - 1, 1)
        paste.SetSourceRegion(sourceRegion)

        destinationIndex = reader.GetOutput().GetLargestPossibleRegion().GetIndex()
        destinationIndex.SetElement(Dimension - 1, i)
        paste.SetDestinationIndex(destinationIndex)

        paste.Update()

        # Fill in the output geometry object
        outputGeometry.SetRadiusCylindricalDetector(
            geometry.GetRadiusCylindricalDetector()
        )
        outputGeometry.AddProjectionInRadians(
            geometry.GetSourceToIsocenterDistances()[src_idx],
            geometry.GetSourceToDetectorDistances()[src_idx],
            geometry.GetGantryAngles()[src_idx],
            geometry.GetProjectionOffsetsX()[src_idx],
            geometry.GetProjectionOffsetsY()[src_idx],
            geometry.GetOutOfPlaneAngles()[src_idx],
            geometry.GetInPlaneAngles()[src_idx],
            geometry.GetSourceOffsetsX()[src_idx],
            geometry.GetSourceOffsetsY()[src_idx],
        )
        outputGeometry.SetCollimationOfLastProjection(
            geometry.GetCollimationUInf()[src_idx],
            geometry.GetCollimationUSup()[src_idx],
            geometry.GetCollimationVInf()[src_idx],
            geometry.GetCollimationVSup()[src_idx],
        )

    # Geometry writer
    if args_info.verbose:
        print(f"Writing geometry information in {args_info.out_geometry}...")
    rtk.write_geometry(outputGeometry, args_info.out_geometry)

    # Write
    if args_info.verbose:
        print(f"Writing selected projections to {args_info.out_proj}...")
    itk.imwrite(paste.GetOutput(), args_info.out_proj)


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
